src/utils/src/ThreadsafeBlockingQueue.c

```diff
@@ -0,0 +1,214 @@
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+
+    CHK(pSafeQueue != NULL, STATUS_NULL_ARG);
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    // 0 timeout semaphore acquire, eventually getting a timeout, this is done to clear
+    // the counting semaphore
+    while ((semaphoreAcquire(pSafeQueue->semaphore, 0) == STATUS_SUCCESS) && !ATOMIC_LOAD_BOOL(&pSafeQueue->terminate))
+        ;
+
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    MUTEX_LOCK(pSafeQueue->mutex);
+    locked = TRUE;
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    CHK_STATUS(stackQueueClear(pSafeQueue->queue, freeData));
+
+    MUTEX_UNLOCK(pSafeQueue->mutex);
+    locked = FALSE;
+
+CleanUp:
+    if (locked) {
+        MUTEX_UNLOCK(pSafeQueue->mutex);
+    }
+
+    return retStatus;
+}
+
+/**
+ * Gets the number of items in the stack/queue
+ */
+STATUS safeBlockingQueueGetCount(PSafeBlockingQueue pSafeQueue, PUINT32 pCount)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+
+    CHK(pSafeQueue != NULL, STATUS_NULL_ARG);
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    MUTEX_LOCK(pSafeQueue->mutex);
+    locked = TRUE;
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    CHK_STATUS(stackQueueGetCount(pSafeQueue->queue, pCount));
+
+    MUTEX_UNLOCK(pSafeQueue->mutex);
+    locked = FALSE;
+
+CleanUp:
+    if (locked) {
+        MUTEX_UNLOCK(pSafeQueue->mutex);
+    }
+
+    return retStatus;
+}
+
+/**
+ * Whether the thread safe blocking queue is empty
+ */
+STATUS safeBlockingQueueIsEmpty(PSafeBlockingQueue pSafeQueue, PBOOL pIsEmpty)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+
+    CHK(pSafeQueue != NULL && pIsEmpty != NULL, STATUS_NULL_ARG);
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    MUTEX_LOCK(pSafeQueue->mutex);
+    locked = TRUE;
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    CHK_STATUS(stackQueueIsEmpty(pSafeQueue->queue, pIsEmpty));
+
+    MUTEX_UNLOCK(pSafeQueue->mutex);
+    locked = FALSE;
+
+CleanUp:
+    if (locked) {
+        MUTEX_UNLOCK(pSafeQueue->mutex);
+    }
+
+    return retStatus;
+}
+
+/**
+ * Enqueues an item in the queue
+ */
+STATUS safeBlockingQueueEnqueue(PSafeBlockingQueue pSafeQueue, UINT64 item)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+
+    CHK(pSafeQueue != NULL, STATUS_NULL_ARG);
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    MUTEX_LOCK(pSafeQueue->mutex);
+    locked = TRUE;
+    CHK(!ATOMIC_LOAD_BOOL(&pSafeQueue->terminate), STATUS_INVALID_OPERATION);
+
+    CHK_STATUS(stackQueueEnqueue(pSafeQueue->queue, item));
+
+    MUTEX_UNLOCK(pSafeQueue->mutex);
+    locked = FALSE;
+
+    CHK_STATUS(semaphoreRelease(pSafeQueue->semaphore));
```

Comment:
Perhaps I am not fully understanding the logic of the blocking queue. I would imagine the semaphore is used to ensure the queue waits until the items are all gone before freeing? If so, wouldn't the semaphore permits be acquired when an item gets enqueued and released when an item gets dequeued?

Reply:
The semaphore acts as a count of items in the queue, and forces operations to block when the queue is empty. A signal with a count of permissions of quick access.

Reply:
OK